File: Example/BPenDemoOC/BPPenCell.m
//
//  BPPenCell.m
//  BPenDemoOC
//
//  Created by Xingfa Zhou on 2023/3/18.
//  Copyright © 2023 bbb. All rights reserved.
//

#import "BPPenCell.h"


@interface BPPenCell()
@property (weak, nonatomic) IBOutlet UILabel* nameLabel;
@property (weak, nonatomic) IBOutlet UILabel* macLabel;
@property (weak, nonatomic) IBOutlet UIButton* connectButton;
@property (weak, nonatomic) IBOutlet UILabel* powerInforLabel;
@property (weak, nonatomic) IBOutlet UISegmentedControl* penSyncDataModeSegement;
@property (weak, nonatomic) IBOutlet UIButton* flashButton;
@property (weak, nonatomic) IBOutlet UILabel* fireworkLabel;
@property (weak, nonatomic) IBOutlet UIButton* fireworkCheckButton;
@property (weak, nonatomic) IBOutlet UIButton* fireworkOTAButton;
@property (weak, nonatomic) IBOutlet UILabel* unSyncDataPercentInforLabel;
@property (weak, nonatomic) IBOutlet UIButton* updateUNSyncDataPercentButton;
@property (weak, nonatomic) IBOutlet UIButton* sleepTimeSetButton;

@property (weak, nonatomic) IBOutlet UIView* backBgView;
@property (weak, nonatomic) IBOutlet UIStackView* stateStack1;
@property (weak, nonatomic) IBOutlet UIStackView* stateStack2;
@property (weak, nonatomic) IBOutlet UIStackView* stateStack3;
@property (weak, nonatomic) IBOutlet UIStackView* stateStack4;

@property (weak, nonatomic) IBOutlet UIButton* startSyncButton;
@property (weak, nonatomic) IBOutlet UIButton* clearFlashDataButton;


@property (strong, nonatomic) BPenModel* pen;
@end

@implementation BPPenCell

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

- (void) configWithPen:(BPenModel*) pen {
    self.pen = pen;
    [self adjustUIForPen:pen];
}

-(void)adjustUIForPen:(BPenModel*) pen {
    self.nameLabel.text = [NSString stringWithFormat:@"name:%@",pen.name];//"name:\(pen.name)"
    self.macLabel.text = [NSString stringWithFormat:@"mac:%@",pen.mac];//"mac:\(pen.mac)"
    if (pen.isConnect) {
        [self.connectButton setTitle:@"断开连接" forState:UIControlStateNormal];
        self.stateStack1.hidden = false;
        self.stateStack2.hidden = false;
        self.stateStack3.hidden = false;
        self.stateStack4.hidden = false;
        self.penSyncDataModeSegement.selectedSegmentIndex = pen.dataSyncMode;
        self.fireworkOTAButton.enabled = pen.canUpdateFirmware;
        self.fireworkLabel.text = [NSString stringWithFormat:@"固件版本:%@",pen.firmwareVersion];//"固件版本:\(pen.firmwareVersion ?? "-")"
        self.fireworkCheckButton.enabled = pen.firmwareVersion.length > 0;// !(pen.firmwareVersion.isEmpty ?? true)
        self.powerInforLabel.text = [NSString stringWithFormat:@"电量:%ld%%",(long)pen.powerPercent];//"电量:\(pen.powerPercent)%"
        self.unSyncDataPercentInforLabel.text = [NSString stringWithFormat:@"已用笔内空间:%ld%%",(long)pen.diskUsedPercent];//"未同步数据占比:\(pen.diskUsedPercent)%"
         
    } else {
       
        [self.connectButton setTitle:pen.isInDFU ? @"更新固件": @"连接设备" forState:UIControlStateNormal];
        self.stateStack1.hidden = true;
        self.stateStack2.hidden = true;
        self.stateStack3.hidden = true;
        self.stateStack4.hidden = true;
    }
}

- (IBAction)connectButtonTouched:(id) sender  {
    if(self.connectOrDisConnectHandel) {
        self.connectOrDisConnectHandel(self.pen);
    }
}

- (IBAction)swithSyncDataModeSegementValueChanged:(id) sender{
    if(self.penSyncDataModeChangeHandel) {
        self.penSyncDataModeChangeHandel(self.pen);
    }
}

- (IBAction)flashButtonTouched:(id) sender{
    if(self.penFlashHandel) {
        self.penFlashHandel(self.pen);
    }
}

- (IBAction)firewareVersionCheckButtonTouched:(id) sender{
    if(self.fireworkCheckHandel) {
        self.fireworkCheckHandel(self.pen);
    }
}


- (IBAction)otaButtonTouched:(id) sender{
    if(self.fireworkOTAHandel) {
        self.fireworkOTAHandel(self.pen);
    }
}

- (IBAction)updateUnSyncDataPercentButtonTouched:(id) sender{
    if(self.updateUnSyncDataPercentHandel) {
        self.updateUnSyncDataPercentHandel(self.pen);
    }
}

- (IBAction)sleeptimeSetButtonTouched:(id) sender{
    if(self.sleepTimeSetHandel) {
        self.sleepTimeSetHandel(self.pen);
    }
}

- (IBAction)startSyncButtonTouched:(id) sender{
    if(self.startSyncHandel) {
        self.startSyncHandel(self.pen);
    }
}

- (IBAction)clearFlashDataButtonTouched:(id) sender{
    if(self.clearFlashDataniuHandel) {
        self.clearFlashDataniuHandel(self.pen);
    }
}
@end

File: Example/BPenDemoSwift/BPPenCell.swift
//
//  BPPenCell.swift
//  BPenPLusDemoSwift
//
//  Created by xingfa on 2021/12/28.
//  Copyright © 2021 bbb. All rights reserved.
//

import UIKit
import BPenPlusSDK

class BPPenCell: UITableViewCell {
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var macLabel: UILabel!
    @IBOutlet weak var connectButton: UIButton!
    @IBOutlet weak var powerInforLabel: UILabel!
    @IBOutlet weak var penSyncDataModeSegement: UISegmentedControl!
    @IBOutlet weak var flashButton: UIButton!
    @IBOutlet weak var fireworkLabel: UILabel!
    @IBOutlet weak var fireworkCheckButton: UIButton!
    @IBOutlet weak var fireworkOTAButton: UIButton!
    @IBOutlet weak var unSyncDataPercentInforLabel: UILabel!
    @IBOutlet weak var updateUNSyncDataPercentButton: UIButton!
    @IBOutlet weak var sleepTimeSetButton: UIButton!
    
    @IBOutlet weak var backBgView: UIView!
    @IBOutlet weak var stateStack1: UIStackView!
    @IBOutlet weak var stateStack2: UIStackView!
    @IBOutlet weak var stateStack3: UIStackView!
    @IBOutlet weak var stateStack4: UIStackView!
    
    @IBOutlet weak var startSyncButton: UIButton!
    @IBOutlet weak var clearFlashDataButton: UIButton!
    
    
    static let reuseIdentifier = "kBPPenCellReuseIdentifier"
    
    var pen: BPenModel?
    var connectOrDisConnectHandel: ((BPenModel) -> Void)?
    var penSyncDataModeChangeHandel: ((BPenModel) -> Void)?
    var penFlashHandel: ((BPenModel) -> Void)?
    var fireworkCheckHandel: ((BPenModel) -> Void)?
    var fireworkOTAHandel: ((BPenModel) -> Void)?
    var updateUnSyncDataPercentHandel: ((BPenModel) -> Void)?
    var sleepTimeSetHandel: ((BPenModel) -> Void)?
    var startSyncHandel: ((BPenModel) -> Void)?
    var clearFlashDataniuHandel: ((BPenModel) -> Void)?
    
    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
        backBgView.layer.borderWidth = 0.5
        backBgView.layer.borderColor = UIColor.gray.cgColor
        backBgView.layer.cornerRadius = 3
        
    }
    
    func configWithPen(_ pen:BPenModel) {
        self.pen = pen
        adjustUIForPen(pen)
    }
    
    func adjustUIForPen(_ pen:BPenModel) {
        nameLabel.text = "name:\(pen.name)"
        macLabel.text = "mac:\(pen.mac)"
        if pen.isConnect {
            connectButton.setTitle("断开连接", for: .normal)
            stateStack1.isHidden = false
            stateStack2.isHidden = false
            stateStack3.isHidden = false
            stateStack4.isHidden = false
            penSyncDataModeSegement.selectedSegmentIndex = pen.dataSyncMode.rawValue
            fireworkOTAButton.isEnabled = pen.canUpdateFirmware
            fireworkLabel.text = "固件版本:\(pen.firmwareVersion ?? "-")"
            fireworkCheckButton.isEnabled = !(pen.firmwareVersion?.isEmpty ?? true)
            powerInforLabel.text = "电量:\(pen.powerPercent)%"
            unSyncDataPercentInforLabel.text = "未同步数据占比:\(pen.diskUsedPercent)%"
        } else {
           
            connectButton.setTitle( pen.isInDFU ? "更新固件": "连接设备", for: .normal)
            stateStack1.isHidden = true
            stateStack2.isHidden = true
            stateStack3.isHidden = true
            stateStack4.isHidden = true
        }
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }
    
    @IBAction func connectButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let connectOrDisConnectHandel = connectOrDisConnectHandel else {
            return
        }
        
        connectOrDisConnectHandel(pen);
    }
    
    @IBAction func swithSyncDataModeSegementValueChanged(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let penSyncDataModeChangeHandel = penSyncDataModeChangeHandel else {
            return
        }
        
        penSyncDataModeChangeHandel(pen);
      //  adjustUIForPen(pen)
    }
    
    @IBAction func flashButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let penFlashHandel = penFlashHandel else {
            return
        }
        
        penFlashHandel(pen);
    }
    
    @IBAction func firewareVersionCheckButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let fireworkCheckHandel = fireworkCheckHandel else {
            return
        }
        
        fireworkCheckHandel(pen);
    }
    
    
    @IBAction func otaButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let fireworkOTAHandel = fireworkOTAHandel else {
            return
        }
        
        fireworkOTAHandel(pen);
    }
    
    @IBAction func updateUnSyncDataPercentButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let updateUNSyncDataPercentHandel = updateUnSyncDataPercentHandel else {
            return
        }
        
        updateUNSyncDataPercentHandel(pen);
    }
    
    @IBAction func sleeptimeSetButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let sleepTimeSetHandel = sleepTimeSetHandel else {
            return
        }
        
        sleepTimeSetHandel(pen);
    }
    
    @IBAction func startSyncButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let startSyncHandel = startSyncHandel else {
            return
        }
        
        startSyncHandel(pen);
    }
    
    @IBAction func clearFlashDataButtonTouched(_ sender: Any) {
        guard let pen = pen else {
            return
        }
        
        guard let clearFlashDataniuHandel = clearFlashDataniuHandel else {
            return
        }
        
        clearFlashDataniuHandel(pen);
        
        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .microseconds(100)) {
            self.updateUnSyncDataPercentButtonTouched(self.updateUNSyncDataPercentButton as Any)
        }
    }
}
